### PR TITLE
Decoupling the aspect generation from the loggable aspect

### PR DESCRIPTION
**Pull request by kpacha**
Adapt the architecture to accept more than one aspect in the same aspect generator

### CODE REVIEW

**Comment by odino** [comment]
hey @kpacha this looks very good, thanks!

@cirpo, @odracci, any feedback?

**Comment by kpacha** [comment]
You are welcome!

I'm porting the GAG package to php and I was looking for a small aop lib like this one. With this refactor, adding more aspects is very easy as you could see at https://github.com/kpacha/php-gag/tree/master/lib/Kpacha/PhpGag/Aspects.

**Comment by kpacha** [comment]
The last two commits were in order to isolate responsibilities and generate more clear classes. With those changes the new class has its methods 'splitted' by stage and every one of those stages starts with a comment. Also, as all is done inside the AspectedMethod or inside the AspectCodeGenerator, the AspectGenerator can focus just on process the annotated aspects ignoring the 'code patterns' to use, the 'formatting facets' like the EOL char and so on.

I know the generated classes are not intended to be human readable but it helps a lot with testing and debugging!

File: src/PUGX/AOP/DependencyInjection/Compiler/Symfony2.php [hunk 1]
```diff
@@ -82,7 +82,13 @@ protected function compileDefinition($id, Definition $definition)
...
             $definition->setClass($pg->getClassName($refClass));
-            $definition->addArgument(new Reference($this->serviceName));
+            if(is_array($this->serviceName)){
+                foreach($this->serviceName as $serviceName){
+                    $definition->addArgument(new Reference($serviceName));
+                }
+            } else {
+                    $definition->addArgument(new Reference($this->serviceName));
+            }
         }
```
Comment (odracci) [inline src/PUGX/AOP/DependencyInjection/Compiler/Symfony2.php:89]: My suggestion is to remove the if statement

``` php
foreach($this->serviceName as $serviceName){
     $definition->addArgument(new Reference($serviceName));
}
```

and put in the constructor

``` php
$this->serviceName = is_array($serviceName) ? $serviceName : array($serviceName);
```

File: src/PUGX/AOP/AspectGenerator.php [hunk 2]
```diff
@@ -76,76 +73,11 @@ public function hasAnnotation(\ReflectionClass $originalClass)
...
-            if ($method->name === '__construct') {
-                $genMethod->addParameter($loaderParam);
-            }
-
-            $generatedClass->setMethod($genMethod);
+        foreach ($this->getMethods($originalClass) as $method) {
+            $generatedClass->setMethod($this->generateMethod($method));
         }
+
+        $generatedClass = $this->prepareAspectInjection($originalClass, $generatedClass);
```
Comment (odracci) [inline src/PUGX/AOP/AspectGenerator.php]: remove the unused variable

**Comment by odracci** [comment]
+1

File: src/PUGX/AOP/DependencyInjection/Compiler/Symfony2.php (hunk 1 above)
Comment (kpacha) [inline src/PUGX/AOP/DependencyInjection/Compiler/Symfony2.php:89]: you're right!

File: src/PUGX/AOP/AspectGenerator.php (hunk 2 above)
Comment (kpacha) [inline src/PUGX/AOP/AspectGenerator.php]: woooooops!

I was trying to return something to the proxy generator in order to chain several aspect generators but, at some point, I changed my mind... ;P

**Comment by kpacha** [comment]
done!

thanks for the feedback, @odracci

**Comment by odino** [comment]
hey @kpacha, are you done with this? Overall looks good to me. Only thing Im keen on is that the docs are updated! :)

@cirpo how does this look to you?

**Comment by lisachenko** [comment]
Can I post a link here to my implementation of Loggable aspect? It's related to this PR )  http://go.aopphp.com/blog/2013/07/21/implementing-logging-aspect-with-doctrine-annotations/

**Comment by odino** [comment]
yes, go for it ;-)

**Comment by kpacha** [comment]
@odino, I think this could do the job for now...

**Comment by kpacha** [comment]
also, I'm waiting to open two new PRs with more minor improvements, like preserving 3rd party annotations and a new 'validation' stage (maybe there are a lot of better names for it, ;P) where you can alter the parameters of the method... but that's another story!

I must wait beacuse I opened this PR from my master branch... yep, shame on me! XD

**Comment by odino** [comment]
Goes on `master`! ;-)

**Pull request merged**